### PR TITLE
REGR: Don't ignore compiled patterns in replace

### LINKED ISSUE #35680: BUG: Different results from Series.replace with compiled regex
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
# Your code here
import re
import pandas as pd

print(pd.__version__)

s = pd.Series(['a', 'aa', 'b'])
regex = {'^a$': 1, '^b$': 2, '^aa$': 3}
print(s.replace(regex, regex=True)) 
print(s.replace({re.compile(k): v for k,v in regex.items()}, regex=True))


```

#### Problem description
These two `.replace` calls are equivalent on pandas `1.0.5`. They are different on `1.1.0` and `master`.

```
1.2.0.dev0+79.g3c87b019b
0    1
1    3
2    2
dtype: int64
0     a
1    aa
2     b
dtype: object


1.0.5
0    1
1    3
2    2
dtype: int64
0    1
1    3
2    2
dtype: int64
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.7.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
machine          : AMD64
processor        : Intel64 Family 6 Model 63 Stepping 2, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.None

pandas           : 1.0.5
numpy            : 1.19.1
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.2.1
setuptools       : 49.3.1.post20200810
Cython           : 0.29.21
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by dsaxton**
- [x] closes #35680
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by dsaxton** [comment]
Currently failing mypy and not quite sure of the fix. https://github.com/pandas-dev/pandas/pull/35697/commits/5006980843a382cd27291e75a8173ec70c00fda6 did not work.

```
mypy --version
mypy 0.730
Performing static analysis using mypy
pandas/core/internals/managers.py:1952: error: Module has no attribute "Pattern"  [attr-defined]
Found 1 error in 1 file (checked 1030 source files)
Performing static analysis using mypy DONE
##[error]Process completed with exit code 1.
```

**Review by arw2019 — COMMENTED** [review]
@dsaxton looked into that CI error you're getting.

xref  https://github.com/python/typing/issues/23 

On my machine adding 
```
from typing import Pattern
```
at the top of `pandas/core/internals/managers.py` and then using plain `Pattern` in the line you modify fixed the `mypy` complaint.

Hope this helps!

**Comment by dsaxton** [comment]
> @dsaxton looked into that CI error you're getting.
> 
> xref [python/typing#23](https://github.com/python/typing/issues/23)
> 
> On my machine adding
> 
> ```
> from typing import Pattern
> ```
> 
> at the top of `pandas/core/internals/managers.py` and then using plain `Pattern` in the line you modify fixed the `mypy` complaint.
> 
> Hope this helps!

Awesome thanks @arw2019

File: pandas/core/internals/managers.py [hunk 1]
```diff
@@ -1949,7 +1959,7 @@ def _check_comparison_types(
     else:
         op = np.vectorize(
             lambda x: bool(re.search(b, x))
-            if isinstance(x, str) and isinstance(b, str)
+            if isinstance(x, str) and isinstance(b, (str, Pattern))
```
Comment (simonjayhawkins) [inline pandas/core/internals/managers.py:1962]: `b` is typed as `Scalar`, where
`Scalar = Union[PythonScalar, PandasScalar]`

but `b` resolves to `Union[builtins.str, builtins.float, Any]` due to unresolved imports and hence mypy is green

but should probably update types and docstring anyway

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @dsaxton generally lgtm

**Comment by simonjayhawkins** [comment]
> Currently failing mypy and not quite sure of the fix. [5006980](https://github.com/pandas-dev/pandas/commit/5006980843a382cd27291e75a8173ec70c00fda6) did not work.
> 
> ```
> mypy --version
> mypy 0.730
> Performing static analysis using mypy
> pandas/core/internals/managers.py:1952: error: Module has no attribute "Pattern"  [attr-defined]
> Found 1 error in 1 file (checked 1030 source files)
> Performing static analysis using mypy DONE
> ##[error]Process completed with exit code 1.
> ```

I think this is a mypy issue that is fixed in later mypy, but can't see a relevant issue on mypy issue tracker

File: pandas/core/internals/managers.py (hunk 1 above)
Comment (dsaxton) [reply pandas/core/internals/managers.py:1962]: Thanks @simonjayhawkins. Would the proper type for b be Union[Scalar, Pattern], or should PythonScalar be extended to include Pattern?

Comment (simonjayhawkins) [reply pandas/core/internals/managers.py:1962]: I would keep PythonScalar unchanged for now.

Comment (dsaxton) [reply pandas/core/internals/managers.py:1965]: Updated. Odd that b is typed as scalar but there's an isinstance check for np.ndarray inside the function, maybe that's not needed.

Comment (simonjayhawkins) [reply pandas/core/internals/managers.py:1965]: the docstring also implies b could be an array. I guess could add ArrayLike while here or leave. either works for me.

Comment (dsaxton) [reply pandas/core/internals/managers.py:1965]: I'll add it. Actually a can also be a scalar so the type shouldn't only be ArrayLike I don't think:
```python
In [17]: b
Out[17]: array(['a', 'a', 'a'], dtype='<U1')

In [18]: _compare_or_regex_search("a", b)
Out[18]: array([ True,  True,  True])
```

Comment (simonjayhawkins) [reply pandas/core/internals/managers.py:1967]: it looks like the types were originally this way when added, see also https://github.com/pandas-dev/pandas/pull/32890#discussion_r406393255, but changed in #33598 

maybe best to keep the changes here minimal for now as this PR is being backported

(and separate PR for ensuring the types are correct and consistent with docstrings.)

File: pandas/core/internals/managers.py [hunk 2]
```diff
@@ -1928,7 +1941,7 @@ def _compare_or_regex_search(
     """
 
     def _check_comparison_types(
-        result: Union[ArrayLike, bool], a: ArrayLike, b: Scalar,
+        result: Union[ArrayLike, bool], a: ArrayLike, b: Union[Scalar, Pattern],
```
Comment (simonjayhawkins) [inline pandas/core/internals/managers.py:1944]: is a `Pattern` ever passed through to `_check_comparison_types`

Comment (simonjayhawkins) [reply pandas/core/internals/managers.py:1944]: I guess it's just for the error message

**Review by simonjayhawkins — APPROVED** [review]
Thanks @dsaxton lgtm

**Comment by simonjayhawkins** [comment]
not a regression, but I guess `regex=True` shouldn't be necessary when compiled regex is passed and if `regex=False` is explicitly passed should error.

**Comment by simonjayhawkins** [comment]
Thanks @dsaxton merging and backporting this as addresses reported regression. followups can be left for 1.2.

**Comment by simonjayhawkins** [comment]
@meeseeksdev backport 1.1.x

**Pull request merged**